### ISSUE TITLE
Endless loop on non-SSH server

### ISSUE THREAD

**Issue by danielzgtg** [ronf/asyncssh#260]
asyncssh doesn't close the connection when connecting to a server that isn't SSH. For example, connecting to a server that outputs `/dev/urandom` and blackholes input will cause asyncssh to get stuck. This causes a denial-of-service on the client.

# Server example

```bash
ncat -l 9999 --send-only --exec "/bin/cat /dev/urandom"
```

# Procedure

```python
import asyncio, asyncssh
asyncio.run(asyncssh.get_server_host_key('127.0.0.1', 9999)
```

# Actual behavior

The `asyncio.run` line blocks forever when pasted into REPL. The Python process stays at 100% CPU.

# Expected behavior

The connection is closed after at most a few seconds. An exception is then thrown.

OpenSSH behaves like this:

```
$ ssh -p 9999 127.0.0.1
kex_exchange_identification: banner line contains invalid characters
```

# Environment

## requirements.txt

```
asyncssh==2.1.0
```

## `python --version`

```
Python 3.8.0
```

## `lsb_release -a`

```
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu Focal Fossa (development branch)
Release:        20.04
Codename:       focal
```

**Comment by ronf** [comment]
Thanks for the report! It's true that AsyncSSH right now will wait forever by default for the "version" line from the server. It also doesn't put an upper bound on the amount of data it will read before a newline. So, if that newline never comes, it could buffer large amounts of data, eating up RAM. Checking RFC 4253, the maximum allowed length of the version string is 255 characters, but I'm not currently enforcing that.

As a mitigation for now, you can add your own timeout around the asyncssh.connect() call using asyncio.wait_for(). If the remote system is feeding large amounts of data, it may still use 100% CPU and some memory for however long a timeout you set, but at least you'll eventually break out and can clean things up.

I'll look into putting some default limits for this.

**Comment by ronf** [comment]
Ok - I've added functionality to help with this in commit d3860e6 in the "develop" branch. It consists of limits on the length of the SSH version line and banner lines before it and on the number of allowed banner lines before the version. So, any server generating a large amount of output that doesn't match an SSH version line will trigger these limits and cause the client to close the connection.

In addition, to deal with servers that don't send a version string or take a long time to do so, I've extended the existing server-side "login_timeout" parameter to also be available on the client side. This timeout actually covers not only the version exchange but also the SSH key exchange and authentications steps. If all of those don't complete within the configured login timeout, the client will drop the connection. This value defaults to two minutes just like on the server side, but it can be adjusted to whatever you like. It can also be disabled by setting it to 0.

**Comment by danielzgtg** [comment]
Thank you. This example seems to produce the expected behavior as of that commit.

**Comment by ronf** [comment]
Thanks for the confirmation! This commit will be included in the next AsyncSSH release.

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.2.0.
